spec/v3/spec.md

```diff
@@ -1,93 +1,372 @@
+
+#### Encoding
+
+To encode a latitude or a longitude into 4 bytes use the following method:
+
+1. Multiply value by 10,000,000
+1. Convert result into little-endian 32-bit signed integer
+
+#### Decoding
+
+To decode a latitude or a longitude from 4 bytes use the following method:
+
+1. Read bytes as a little-endian 32-bit signed integer
+1. Divide read value by 10,000,000
+
+## 4 Directories
+
+A directory is simply a list of entries. Each entry describes either where a specific tile can be found in the _tile data section_, or where a leaf directory can be found in the _leaf directories section_.  
+
+The number of entries in the root directory and in the leaf directories is left to the implementation and can vary drastically depending on what the writer has optimized for (cost, bandwidth, latency etc.).  
+However, the size of the header plus the compressed size of the root directory must not exceed 16384 bytes to allow latency-optimized clients to retrieve the root directory in its entirety. Therefore, the **maximum compressed size of the root directory is 16257 bytes** (16384 bytes - 127 bytes). A sophisticated writer might need several attempts to optimize this.
+
+### 4.1 Directory Entries
+
+Each directory entry consists of the following properties:
+- Tile ID
+- Offset
+- Length
+- Run-Length
+
+#### Tile-ID
+Specifies the ID of the tile / the first tile in the leaf directory. 
+
+The Tile-ID corresponds to a cumulative position on the series of [Hilbert curves](https://wikipedia.org/wiki/Hilbert_curve) starting at zoom level 0.
+
+|Z|X|Y|TileID|
+|--:|--:|--:|--:|
+|0|0|0|0|
+|1|0|0|1|
+|1|0|1|2|
+|1|1|0|3|
+|1|1|1|4|
+|2|0|0|5|
+|...
+|12|3423|1763|19078479|
+
+#### Offset
+Specifies the offset of the first byte of the tile or leaf directory. This address offset is relative to the first byte of the _tile data section_ for tile-entries, and relative to the first byte of the _leaf directories section_ for leaf-directory-entries.
+
+#### Length
+Specifies the number of bytes reserved for this tile or leaf directory. This size always indicates the compressed size, if the tile or leaf directory is compressed.
+
+#### Run-Length
+Specifies the number of tiles for which this entry is valid. A run length of `0` means that this entry is for a leaf directory and not for a tile.
+
+#### Examples
+|Tile-ID|Offset|Length|Run-Length|Description|
+|--:|--:|--:|--:|:--|
+|`5`|`1337`|`42`|`1`|Tile 5 is located at bytes 1337-1378 of the _tile data section_|
+|`5`|`1337`|`42`|`3`|Tile 5, 6 and 7 are located at bytes 1337-1378 of the _tile data section_|
+|`5`|`1337`|`42`|`0`|A leaf directory whose first tile has an ID of 5 is located at byte 1337-1378 of the _leaf directories section_|
+
+### 4.2 Encoding
+A directory can only be encoded in its entirety. It is not possible to encode a single directory entry by itself. 
```

Comment:
@bdon in https://github.com/protomaps/PMTiles/issues/107#issuecomment-1535056512:

> What implies this? An archive that addresses a single tile should be valid, AFAIK.

Reply:
The values you encode for an entry always depend on the entry that came before that entry. That's why I said that entries can't be encoded by itself.

Yes an archive that addresses a single tile should be valid, but that would have to have a root directory with one entry. At least you would to have to prepend a `1` to indicate the number of entries, that are contained in the directory and at that point it's a directory and not a single entry.

Reply:
> At least you would to have to prepend a 1 to indicate the number of entries, that are contained in the directory and at that point it's a directory and not a single entry.

OK, I see what you mean here. I meant to say that a directory with length 1 is valid; any directory with length > 0 is valid. So I think 4.2 should simply say "A directory must have at least one entry".

Reply:
I'm happy to add something along the lines of that, although it's already mentioned in another chapter.

But I still think we should keep something along the lines of _"A directory can only be encoded in its entirety"_, because I remember that wasn't really clear to me when I first started implementing PMTiles. What I wanted to know then was how the entries are encoded / decoded, but that's just the wrong starting point, since they are always interleaved.